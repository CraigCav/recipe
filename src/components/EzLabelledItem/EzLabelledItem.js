import React from 'react';
import PropTypes from 'prop-types';
import styled from 'react-emotion';
import {base, position} from './EzLabelledItem.styles';

const LabelWrapper = styled.div(base, position);

/**
 * Provides a label that can appear in one of several positions relative to its associated content.
 * Typically used to label an interactive element or a short string of text.
 */
const EzLabelledItem = ({children, position: labelPosition, size, title}) => (
  <div>
    {labelPosition === 'bottom' && children}
    <LabelWrapper position={labelPosition} size={size}>
      {title}
    </LabelWrapper>
    {['top', 'left'].includes(labelPosition) && children}
  </div>
);

EzLabelledItem.propTypes = {
  /**
   * The content to render in conjunction with the label
   */
  children: PropTypes.node.isRequired,
  /**
   * Determines the position of the label relative to its child content
   */
  position: PropTypes.oneOf(['top', 'bottom', 'left']).isRequired,
  /**
   * Determines the size of the label text
   */
  size: PropTypes.oneOf(['normal', 'small']),
  /**
   * The text for label itself
   */
  title: PropTypes.node.isRequired,
};

/**
 * defaultProps
 * @property {string} size - uses defaults to the base font size.
 */
EzLabelledItem.defaultProps = {size: 'normal'};

EzLabelledItem.displayName = 'EzLabelledItem';

/**
 * @component
 */
export default EzLabelledItem;
