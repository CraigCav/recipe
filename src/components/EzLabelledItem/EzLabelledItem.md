---
name: EzLabelledItem
title: Labelled Item
path: '/components/ez-labelled-item'
---

Provides a label that can appear in one of several positions relative to its associated content. Typically used to label an interactive element or a short string of text.

---

## Best Practices

Labelled items should:

- Be used to label a static piece of text
- Be used to label a control such as a Dropdown

Labelled items should not:

- Be used to label more than one child (unless being used with a single child that is a collection of multiple items, such as a Segmented Control)

---

## Content guidelines

Labelled items should

#### Do

- Use size="normal" with any position value
- Use size="small" with position="top"
- Use title case for the label title

#### Donâ€™t

- Don't use size="small" with position="left" or position="right"

---

## Examples

### Top label

A label that appears above its content.

```jsx
<EzLabelledItem position="top" size="normal" title="Top Label">
  Some text
</EzLabelledItem>
```

### Top label (small)

This is alternative style of top label.

```jsx
<EzLabelledItem position="top" size="small" title="Small Label">
  Some text
</EzLabelledItem>
```

### Left label

A label that appears next to its content.

```jsx
<EzLabelledItem position="left" size="normal" title="Left Label">
  Some text
</EzLabelledItem>
```

### Bottom label

A label that appears below its content.

```jsx
<EzLabelledItem position="bottom" size="normal" title="Bottom Label">
  Some text
</EzLabelledItem>
```
