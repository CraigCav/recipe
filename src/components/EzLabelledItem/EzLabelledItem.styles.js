import {css} from 'react-emotion';
import {variants} from '../../styles/';

export const base = ({theme}) => css`
  color: ${theme.colors.blueGrays[100]};
`;

const normalSize = ({theme}) => css`
  font-size: ${theme.fontSizes.normal};
  font-weight: ${theme.fontWeights.base};
`;

const top = ({size, theme}) => css`
  font-size: ${theme.fontSizes[size]};
  font-weight: ${size === 'small' && theme.fontWeights.medium};
  margin-bottom: ${theme.spacing.xs2};
`;

const bottom = props => css`
  ${normalSize(props)};
  margin-top: ${props.theme.spacing.xs2};
`;

const left = props => css`
  ${normalSize(props)};
  display: inline;
  margin-right: ${props.theme.spacing.sm};
`;

export const position = variants('position', {
  top,
  bottom,
  left,
});
