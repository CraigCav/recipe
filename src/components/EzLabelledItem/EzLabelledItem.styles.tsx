import {css} from 'react-emotion';
import variant from 'styled-component-variant';

export const base = ({theme}) => css`
  color: ${theme.colors.grays[600]};
`;

const normalSize = ({theme}) => css`
  font-size: ${theme.fontSizes[300]};
  font-weight: ${theme.fontWeights.normal};
`;

const top = ({size, theme}) => css`
  font-size: ${size === 'small' && theme.fontSizes[200]};
  font-weight: ${size === 'small' && theme.fontWeights.bold};
  margin-bottom: ${theme.spacing.xs2};
`;

const bottom = props => css`
  ${normalSize(props)};
  margin-top: ${props.theme.spacing.xs2};
`;

const left = props => css`
  ${normalSize(props)};
  display: inline;
  margin-right: ${props.theme.spacing.sm};
`;

export const position = variant('position', {
  top,
  bottom,
  left,
});
