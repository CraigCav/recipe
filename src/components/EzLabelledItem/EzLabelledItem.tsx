import React from 'react';
import PropTypes from 'prop-types';
import styled from '../../themes/styled';

import {base, position} from './EzLabelledItem.styles';

type LabelWrapperProps = {
  position: string;
  size: string;
};
const LabelWrapper = styled.div<LabelWrapperProps>(base, position);

/**
 * Provides a label that can appear in one of several positions relative to its associated content.
 * Typically used to label an interactive element or a short string of text.
 */
type ItemProps = {
  children: any;
  position?: string;
  size?: string;
  title: string;
};

const EzLabelledItem: React.SFC<ItemProps> = ({children, position: labelPosition, size, title}) => (
  <div>
    {labelPosition === 'bottom' && children}
    <LabelWrapper position={labelPosition} size={size}>
      {title}
    </LabelWrapper>
    {['top', 'left'].includes(labelPosition) && children}
  </div>
);

/**
 * defaultProps
 * @property {string} size - uses defaults to the base font size.
 */
EzLabelledItem.defaultProps = {size: 'normal'};

EzLabelledItem.displayName = 'EzLabelledItem';

/**
 * @component
 */
export default EzLabelledItem;
