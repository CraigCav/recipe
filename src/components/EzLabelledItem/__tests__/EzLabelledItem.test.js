import React from 'react';
import EzLabelledItem from '../EzLabelledItem';

describe('EzLabelledItem', () => {
  describe('position', () => {
    describe('top', () => {
      it('renders the label above the content', () => {
        const actual = render(
          <EzLabelledItem position="top" title="Top Label">
            Some text
          </EzLabelledItem>
        );
        expect(actual.find(':first-child').text()).toContain('Top Label');
      });

      it('renders with a small label', () => {
        const actual = render(
          <EzLabelledItem position="top" size="small" title="Top Label">
            Some text
          </EzLabelledItem>
        );
        expect(actual).toMatchSnapshot();
      });
    });

    describe('left', () => {
      it('renders the label to the left of the content', () => {
        const actual = render(
          <EzLabelledItem position="left" title="Left Label">
            Some text
          </EzLabelledItem>
        );
        expect(actual.find(':first-child').text()).toContain('Left Label');
        expect(actual).toMatchSnapshot();
      });
    });
  });

  it('renders the label below the content', () => {
    const actual = render(
      <EzLabelledItem position="bottom" title="Bottom Label">
        Some text
      </EzLabelledItem>
    );
    expect(actual.find(':last-child').text()).toContain('Bottom Label');
  });

  /**
   * Accessibility tests.
   */
  it('should meet accessibility guidelines', async () => {
    const wrapper = renderToHtml(
      <EzLabelledItem position="top" title="Top Label">
        Some text
      </EzLabelledItem>
    );
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });
});
