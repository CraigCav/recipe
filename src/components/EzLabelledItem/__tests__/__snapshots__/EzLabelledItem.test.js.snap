// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EzLabelledItem position left renders the label to the left of the content 1`] = `
.recipe-0 {
  color: #8B99A6;
  font-size: 1rem;
  font-weight: 400;
  display: inline;
  margin-right: 0.8571428571428571rem;
}

<div>
  <div
    class="recipe-0 recipe-1"
  >
    Left Label
  </div>
  Some text
</div>
`;

exports[`EzLabelledItem position top renders with a small label 1`] = `
.recipe-0 {
  color: #8B99A6;
  font-size: 0.85rem;
  font-weight: 700;
  margin-bottom: 0.2857142857142857rem;
}

<div>
  <div
    class="recipe-0 recipe-1"
  >
    Top Label
  </div>
  Some text
</div>
`;
